File: DP/src/main/java/Solution122.java
public class Solution122 {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int[][] opt = new int[n][2];

        opt[0][0] = 0;
        opt[0][1] = -prices[0];

        for (int i = 1; i < n; i++) {
            opt[i][0] = Math.max(opt[i - 1][0], opt[i - 1][1] + prices[i]);
            opt[i][1] = Math.max(opt[i - 1][1], opt[i - 1][0] - prices[i]);
        }

        return opt[n - 1][0];
    }
}

File: DP/src/main/python/solution122.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        opt = [[float("-inf"), float("-inf")] for i in range(n)]

        opt[0][0] = 0
        opt[0][1] = -prices[0]

        for i in range(1, n):
            opt[i][0] = max(opt[i - 1][0], opt[i - 1][1] + prices[i])

            opt[i][1] = max(opt[i - 1][0] - prices[i], opt[i - 1][1])

        # print(opt)
        return opt[n - 1][0]
